freeglut/freeglut/src/x11/fg_main_x11.c

```diff
@@ -934,46 +1227,71 @@ void fgPlatformProcessSingleEvent ( void )
                 keyboard_ud = FETCH_USER_DATA_WCB( *window, KeyboardUp );
                 special_ud  = FETCH_USER_DATA_WCB( *window, SpecialUp  );
             }
 
+
             /* Is there a keyboard/special callback hooked for this window? */
-            if( keyboard_cb || special_cb )
+            if (keyboard_ext_cb || keyboard_cb || special_cb)
             {
-                XComposeStatus composeStatus;
-                char asciiCode[ 32 ];
+                static XComposeStatus composeStatus = { 0 }; /* keep state across invocations */
+                XIC ic = window->Window.pContext.IC;
+                Status status;
+                char buf[32], *utf8 = buf;
                 KeySym keySym;
-                int len;
+                int i, c, len;
 
                 /* Check for the ASCII/KeySym codes associated with the event: */
-                len = XLookupString( &event.xkey, asciiCode, sizeof(asciiCode),
-                                     &keySym, &composeStatus
-                );
+                if (ic)
+                {
+                    len = Xutf8LookupString(ic, &event.xkey, buf, sizeof buf, &keySym, &status);
+                    if (status == XBufferOverflow)
+                    {
+                        utf8 = malloc(len);
+                        len = Xutf8LookupString(ic, &event.xkey, utf8, len, &keySym, &status);
+                    }
+                }
+                else
+                {
+                    len = XLookupString(&event.xkey, buf, sizeof buf, &keySym, &composeStatus);
+                }
 
-                /* GLUT API tells us to have two separate callbacks... */
-                if( len > 0 )
+                /* GLUT API tells us to have three separate callbacks... */
+                if (len > 0)
                 {
-                    /* ...one for the ASCII translateable keypresses... */
-                    if( keyboard_cb )
+                    fgSetWindow(window);
+                    fgState.Modifiers = fgPlatformGetModifiers(event.xkey.state);
+
+                    i = 0;
+                    while (i < len)
                     {
-                        fgSetWindow( window );
-                        fgState.Modifiers = fgPlatformGetModifiers( event.xkey.state );
-                        keyboard_cb( asciiCode[ 0 ],
-                                     event.xkey.x, event.xkey.y,
-                                     keyboard_ud
-                        );
-                        fgState.Modifiers = INVALID_MODIFIERS;
+                        i += chartorune(&c, utf8 + i);
+
+                        /* ...one for the Unicode translateable keypresses... */
+                        if (keyboard_ext_cb)
+                            keyboard_ext_cb(c, event.xkey.x, event.xkey.y, keyboard_ext_ud);
+
+                        /* ...one for the ASCII translateable keypresses... */
+                        if (keyboard_cb)
+                            if (c < 128)
```

Comment:
Same comment as before: Will this break existing GLUT functionality?

I get it would only be one part of the whole Unicode, but I don't know who might be relying on that value.